macros/shared/src/main/scala-2.x/zio/macros/AccessibleMMacroBase.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2019-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.macros
+
+import scala.reflect.macros.whitebox
+
+import com.github.ghik.silencer.silent
+
+private[macros] abstract class AccessibleMMacroBase(override val c: whitebox.Context) extends AccessibleMacroBase(c) {
+
+  import c.universe._
+
+  protected def expectedTypeParams: Long
+
+  protected def aliases: Seq[Tree]
+
+  private val `_` : TypeName = TypeName("_")
+
+  @silent("pattern var [^\\s]+ in method unapply is never used")
+  protected val tp: Tree = c.prefix.tree match {
+    case q"new ${macroNm: Ident}[$typeParam]" if macroNm.name == TypeName(macroName) =>
+      typeParam
+    case _ =>
+      abort("could not unquote annotation")
+  }
+
+  protected val tpTpe: Type = {
+    val res = c.typecheck(tq"$tp", c.TYPEmode).tpe
+    if (types.keySet.contains(res))
+      res
+    else
+      abort(s"unsupported type constructor $res. Supported constructors: ${types.keySet}")
+  }
+
+  private lazy val types: Map[Type, Tree] =
+    aliases.map(a => c.typecheck(tq"$a", c.TYPEmode).tpe -> a).toMap
+
+  @silent("pattern var [^\\s]+ in method unapply is never used")
+  protected def macroApply(annottees: Seq[c.Tree]): MacroApply = new MacroApply(annottees) {
+
+    private val typeParamToInject = {
+      val candidates =
+        moduleInfo.serviceTypeParams.filter { tp =>
+          tp.tparams.forall(_.name == `_`) && tp.tparams.size == expectedTypeParams
```

Comment:
> tp.tparams.forall(_.name == `_`)

That's an odd check. It's legal to name type parameters in Scala, as in `trait Service[F[e, a]]`

Reply:
Yeah, it was really too strict a requirement. Fixed.